Pydantic Transformer V2

Future-Outlier opened the pull request:
## Tracking issue
https://github.com/flyteorg/flyte/issues/5033
https://github.com/flyteorg/flyte/issues/5318

## How to test it by others
1. git clone https://github.com/flyteorg/flytekit
2. gh pr checkout 2792
3. make setup-global-uv
4. cd plugins/flytekit-pydantic-v2 && pip install -e .
5. test a workflow example

## Not Sure
Which pydantic version should we use as the lower bound?

## This case will fail in the Flyte Console
```python
@dataclass
class DC:
    a: Union[bool, str, int]
    b: Union[bool, str, int]

@task(container_image=image)
def add(dc1: DC, dc2: DC) -> Union[bool, int, str]:
    return dc1.a + dc2.b  # type: ignore

# input from flyte console to generate generic protobuf struct
# "{\"a\": 1, \"b\": 2}",
@workflow
def wf(dc: DC) -> Union[bool, int, str]:
    return add(dc1=dc, dc2=dc)
```

## file tree structure
1. The file tree structure is the same as [flytekit-pydantic](https://github.com/flyteorg/flytekit/tree/master/plugins/flytekit-pydantic)

## Why didn't integrate with pydantic v1 BaseModel? (make you run v1 and v2 BaseModel at the same time together)
This is an issue from pydantic.
https://github.com/pydantic/pydantic/issues/9919
If this is fixed, then we can support both pydantic v1 and v2 at the same time.

story:
Kevin and I wanted to support v1 and v2 at the same time before, but after knowing that this would take lots of time, we asked Ketan for advice, then he said that if users want it, then we can try to support it or tell users to support it.

## Why are the changes needed?
1. why from_generic_idl?
flyteconsole input to handle flytetypes.

when handling the input below, and attribute access to a flyte type, we need to teach flyte types how to convert a protobuf struct to flyte types.
Take FlyteFile as an example.

lifecycle
```zsh
json str -> protobuf struct -> attribute access flyte type -> send to downstream input
```
```python
class DC(BaseModel):
    ff: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))

@workflow
def wf(dc: DC) -> DC:
    t_ff(dc.ff)
    return t_args(dc=dc)

# console input: {"ff":{"path":"s3://my-s3-bucket/example.txt"}}
```
<img width="524" alt="image" src="https://github.com/user-attachments/assets/8be2d2db-363a-4a67-ac30-75c62311fa76">
<img width="1157" alt="image" src="https://github.com/user-attachments/assets/06d40929-ec21-4514-87d6-83a226450e6a">

2. why _check_and_covert_int in the int transformer?
flyteconsole input to handle float issue.
It will be needed when in the following example.
```
json str -> protbuf struct -> attribute access and get float(due to javascript problem) -> convert float to int in flytekit
```

```python
class DC(BaseModel):
    a: int = -1

@workflow
def wf(dc: DC):
    t_int(input_int=dc.a)
```

3. why basemodel -> json str -> dict obj -> msgpack bytes?
For enum class.

I've tried basemodel -> dict obj -> msgpack bytes first.
To make this happen, you need to call the function `BaseModel.model_dump`, but this function can't interpret `Enum`.
However, `BaseModel.model_dump_json` can.

4. What are `@model_serializer` and `@model_validator(mode="after")`?
You can understand them as `_serialize` and `_deserialize` in FlyteTypes, which use `SerializableType` to customize the serialize/deserialize behavior for flyte types.

Related PRs: https://github.com/flyteorg/flytekit/pull/2554

## What changes were proposed in this pull request?
- [x] attribute access (primitives and flyte types) (datetime not sure)
- [x] flyte types
- [x] nested cases
- [x] dataclasses.dataclass in pydantic.BaseModel
- [x] pydantic.dataclass in pydantic.BaseModel
- [x] pydantic.BaseModel in pydantic.BaseModel

note: we don't support pydantic BaseModel has a dataclass with FlyteTypes.
We support pydantic BaseModel has a dataclass with primitive types.
```python
@dataclass
class dc:
    ff: FlyteFile

class DC(BaseModel):
    inner_dc: dc

# This is not supported
# ============================
@dataclass
class dc:
    a: int

class DC(BaseModel):
    inner_dc: dc

# This is supported
# ============================
```

## How was this patch tested?
Example code.
(nested cases, flyte types, and attribute access.)
```python
from pydantic import BaseModel, Field
from typing import Dict, List, Optional

from flytekit.types.schema import FlyteSchema
from flytekit.types.structured import StructuredDataset
from flytekit.types.file import FlyteFile
from flytekit.types.directory import FlyteDirectory
from flytekit import task, workflow, ImageSpec, kwtypes
from enum import Enum
import os
import pandas as pd

flytekit_hash = "fb82dd521615039f626c78489b2e83259d7db2a5"
flytekit = f"git+https://github.com/flyteorg/flytekit.git@{flytekit_hash}"
pydantic_plugin = f"git+https://github.com/flyteorg/flytekit.git@{flytekit_hash}#subdirectory=plugins/flytekit-pydantic-v2"

# Define custom image for the task
image = ImageSpec(packages=[
                            flytekit,
                            pydantic_plugin,
                            "pandas",
                            "pyarrow"],
                            apt_packages=["git"],
                            registry="localhost:30000",
                         )

class Status(Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"

class InnerBM(BaseModel):
    a: int = -1
    b: float = 2.1
    c: str = "Hello, Flyte"
    d: bool = False
    e: List[int] = Field(default_factory=lambda: [0, 1, 2, -1, -2])
    f: List[FlyteFile] = Field(default_factory=lambda: [FlyteFile("s3://my-s3-bucket/example.txt")])
    g: List[List[int]] = Field(default_factory=lambda: [[0], [1], [-1]])
    h: List[Dict[int, bool]] = Field(default_factory=lambda: [{0: False}, {1: True}, {-1: True}])
    i: Dict[int, bool] = Field(default_factory=lambda: {0: False, 1: True, -1: False})
    j: Dict[int, FlyteFile] = Field(default_factory=lambda: {0: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             1: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             -1: FlyteFile("s3://my-s3-bucket/example.txt")})
    k: Dict[int, List[int]] = Field(default_factory=lambda: {0: [0, 1, -1]})
    l: Dict[int, Dict[int, int]] = Field(default_factory=lambda: {1: {-1: 0}})
    m: dict = Field(default_factory=lambda: {"key": "value"})
    n: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))
    o: FlyteDirectory = Field(default_factory=lambda: FlyteDirectory("s3://my-s3-bucket/s3_flyte_dir"))
    enum_status: Status = Status.PENDING
    sd: StructuredDataset = Field(default_factory=lambda: StructuredDataset(
        uri="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/7f31035fdf92510e40ee9340f9e5bf34",
        file_format="parquet"))
    fsc: FlyteSchema = Field(default_factory=lambda: FlyteSchema(
        remote_path="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/ab3aef21302d0529daef8c43825c3fdf"))

class BM(BaseModel):
    a: int = -1
    b: float = 2.1
    c: str = "Hello, Flyte"
    d: bool = False
    e: List[int] = Field(default_factory=lambda: [0, 1, 2, -1, -2])
    f: List[FlyteFile] = Field(default_factory=lambda: [FlyteFile("s3://my-s3-bucket/example.txt")])
    g: List[List[int]] = Field(default_factory=lambda: [[0], [1], [-1]])
    h: List[Dict[int, bool]] = Field(default_factory=lambda: [{0: False}, {1: True}, {-1: True}])
    i: Dict[int, bool] = Field(default_factory=lambda: {0: False, 1: True, -1: False})
    j: Dict[int, FlyteFile] = Field(default_factory=lambda: {0: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             1: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             -1: FlyteFile("s3://my-s3-bucket/example.txt")})
    k: Dict[int, List[int]] = Field(default_factory=lambda: {0: [0, 1, -1]})
    l: Dict[int, Dict[int, int]] = Field(default_factory=lambda: {1: {-1: 0}})
    m: dict = Field(default_factory=lambda: {"key": "value"})
    n: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))
    o: FlyteDirectory = Field(default_factory=lambda: FlyteDirectory("s3://my-s3-bucket/s3_flyte_dir"))
    inner_bm: InnerBM = Field(default_factory=lambda: InnerBM())
    enum_status: Status = Status.PENDING
    sd: StructuredDataset = Field(default_factory=lambda: StructuredDataset(
        uri="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/7f31035fdf92510e40ee9340f9e5bf34",
        file_format="parquet"))
    fsc: FlyteSchema = Field(default_factory=lambda: FlyteSchema(remote_path="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/ab3aef21302d0529daef8c43825c3fdf"))

@task(container_image=image)
def t_bm(bm: BM) -> BM:
    return bm

@task(container_image=image)
def t_inner(inner_bm: InnerBM):
    assert isinstance(inner_bm, InnerBM)

    expected_file_content = "Default content"

    # f: List[FlyteFile]
    for ff in inner_bm.f:
        assert isinstance(ff, FlyteFile)
        with open(ff, "r") as f:
            assert f.read() == expected_file_content
    # j: Dict[int, FlyteFile]
    for _, ff in inner_bm.j.items():
        assert isinstance(ff, FlyteFile)
        with open(ff, "r") as f:
            assert f.read() == expected_file_content
    # n: FlyteFile
    assert isinstance(inner_bm.n, FlyteFile)
    with open(inner_bm.n, "r") as f:
        assert f.read() == expected_file_content
    # o: FlyteDirectory
    assert isinstance(inner_bm.o, FlyteDirectory)
    assert not inner_bm.o.downloaded
    with open(os.path.join(inner_bm.o, "example.txt"), "r") as fh:
        assert fh.read() == expected_file_content
    assert inner_bm.o.downloaded
    print("Test InnerBM Successfully Passed")
    # enum: Status
    assert inner_bm.enum_status == Status.PENDING


@task(container_image=image)
def t_test_all_attributes(a: int, b: float, c: str, d: bool, e: List[int], f: List[FlyteFile], g: List[List[int]],
                          h: List[Dict[int, bool]], i: Dict[int, bool], j: Dict[int, FlyteFile],
                          k: Dict[int, List[int]], l: Dict[int, Dict[int, int]], m: dict,
                          n: FlyteFile, o: FlyteDirectory,
                          enum_status: Status,
                          sd: StructuredDataset,
                          fsc: FlyteSchema,
                          ):
    # Strict type checks for simple types
    assert isinstance(a, int), f"a is not int, it's {type(a)}"
    assert a == -1
    assert isinstance(b, float), f"b is not float, it's {type(b)}"
    assert isinstance(c, str), f"c is not str, it's {type(c)}"
    assert isinstance(d, bool), f"d is not bool, it's {type(d)}"

    # Strict type checks for List[int]
    assert isinstance(e, list) and all(isinstance(i, int) for i in e), "e is not List[int]"

    # Strict type checks for List[FlyteFile]
    assert isinstance(f, list) and all(isinstance(i, FlyteFile) for i in f), "f is not List[FlyteFile]"

    # Strict type checks for List[List[int]]
    assert isinstance(g, list) and all(
        isinstance(i, list) and all(isinstance(j, int) for j in i) for i in g), "g is not List[List[int]]"

    # Strict type checks for List[Dict[int, bool]]
    assert isinstance(h, list) and all(
        isinstance(i, dict) and all(isinstance(k, int) and isinstance(v, bool) for k, v in i.items()) for i in h
    ), "h is not List[Dict[int, bool]]"

    # Strict type checks for Dict[int, bool]
    assert isinstance(i, dict) and all(
        isinstance(k, int) and isinstance(v, bool) for k, v in i.items()), "i is not Dict[int, bool]"

    # Strict type checks for Dict[int, FlyteFile]
    assert isinstance(j, dict) and all(
        isinstance(k, int) and isinstance(v, FlyteFile) for k, v in j.items()), "j is not Dict[int, FlyteFile]"

    # Strict type checks for Dict[int, List[int]]
    assert isinstance(k, dict) and all(
        isinstance(k, int) and isinstance(v, list) and all(isinstance(i, int) for i in v) for k, v in
        k.items()), "k is not Dict[int, List[int]]"

    # Strict type checks for Dict[int, Dict[int, int]]
    assert isinstance(l, dict) and all(
        isinstance(k, int) and isinstance(v, dict) and all(
            isinstance(sub_k, int) and isinstance(sub_v, int) for sub_k, sub_v in v.items())
        for k, v in l.items()), "l is not Dict[int, Dict[int, int]]"

    # Strict type check for a generic dict
    assert isinstance(m, dict), "m is not dict"

    # Strict type check for FlyteFile
    assert isinstance(n, FlyteFile), "n is not FlyteFile"

    # Strict type check for FlyteDirectory
    assert isinstance(o, FlyteDirectory), "o is not FlyteDirectory"

    # # Strict type check for Enum
    assert isinstance(enum_status, Status), "enum_status is not Status"

    assert isinstance(sd, StructuredDataset), "sd is not StructuredDataset"
    print("sd:", sd.open(pd.DataFrame).all())

    assert isinstance(fsc, FlyteSchema), "fsc is not FlyteSchema"
    print("fsc: ", fsc.open().all())

    print("All attributes passed strict type checks.")


@workflow
def wf(bm: BM):
    t_bm(bm=bm)
    t_inner(inner_bm=bm.inner_bm)
    t_test_all_attributes(a=bm.a, b=bm.b, c=bm.c,
                          d=bm.d, e=bm.e, f=bm.f,
                          g=bm.g, h=bm.h, i=bm.i,
                          j=bm.j, k=bm.k, l=bm.l,
                          m=bm.m, n=bm.n, o=bm.o,
                          enum_status=bm.enum_status,
                          sd=bm.sd,
                          fsc=bm.fsc,
                          )

    t_test_all_attributes(a=bm.inner_bm.a, b=bm.inner_bm.b, c=bm.inner_bm.c,
                          d=bm.inner_bm.d, e=bm.inner_bm.e, f=bm.inner_bm.f,
                          g=bm.inner_bm.g, h=bm.inner_bm.h, i=bm.inner_bm.i,
                          j=bm.inner_bm.j, k=bm.inner_bm.k, l=bm.inner_bm.l,
                          m=bm.inner_bm.m, n=bm.inner_bm.n, o=bm.inner_bm.o,
                          enum_status=bm.inner_bm.enum_status,
                          sd=bm.inner_bm.sd,
                          fsc=bm.inner_bm.fsc,
                          )

if __name__ == "__main__":
    from flytekit.clis.sdk_in_container import pyflyte
    from click.testing import CliRunner

    runner = CliRunner()
    path = os.path.realpath(__file__)
    input_val = BM().model_dump_json()
    print(input_val)
    result = runner.invoke(pyflyte.main,
                           ["run", path, "wf", "--bm", input_val])
    print("Local Execution: ", result.output)

    result = runner.invoke(pyflyte.main,
                           ["run", "--remote", path, "wf", "--bm", input_val])
    print("Remote Execution: ", result.output)
```

### Setup process
local and remote execution.
ImageSpec for the docker image.

### Screenshots
- local execution
<img width="1031" alt="image" src="https://github.com/user-attachments/assets/677015d0-2f94-4e91-8301-f244d596ba23">
<img width="208" alt="image" src="https://github.com/user-attachments/assets/84c468bf-e1b6-49a5-92ed-79632f289d77">

- remote execution
<img width="1168" alt="image" src="https://github.com/user-attachments/assets/67362674-ad6a-43c2-93ea-6d756035497a">

- remote execution from flyte console input
<table>
  <tr>
    <td><img width="443" alt="image" src="https://github.com/user-attachments/assets/d9dcfec8-709f-4cf2-83e0-1deddd5a8d0c"></td>
    <td><img width="1163" alt="image" src="https://github.com/user-attachments/assets/930ca4bf-b868-4da5-8bed-5d015f10b93c"></td>
  </tr>
</table>


## Check all the applicable boxes <!-- Follow the above conventions to check the box -->

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

Future-Outlier commented on flytekit/core/type_engine.py:
```diff
@@ -2052,6 +2052,17 @@ def _check_and_covert_float(lv: Literal) -> float:
+    if lv.scalar.primitive.integer is not None:
+        return lv.scalar.primitive.integer
+
+    if lv.scalar.primitive.float_value is not None:
+        logger.info(f"Converting literal float {lv.scalar.primitive.float_value} to int, might have precision loss.")
+        return int(lv.scalar.primitive.float_value)
```
This is for cases when you input from flyte console, and you use attribute access directly, you have to the `float` to `int`.
Since javascript has only `number`, it can't tell the difference between int and float.

```python
class TrainConfig(BaseModel):
    lr: float = 1e-3
    batch_size: int = 32

@workflow
def wf(cfg: TrainConfig) -> TrainConfig:
    return t_args(a=cfg.lr, batch_size=cfg.batch_size)
```

Future-Outlier replied:
YES, the attribute access works well, it's because javascript pass float to golang, and golang pass float to python.

Future-Outlier replied:
> this should only be a javascript problem. attribute access should work since msgpack preserves float/int even in attribute access correct?

Yes, but when you are accessing a simple type, you have to change the behavior of `SimpleTransformer`.

For Pydantic Transformer, we will use `strict=False` as argument to convert it to right type.
```python
    def from_binary_idl(self, binary_idl_object: Binary, expected_python_type: Type[BaseModel]) -> BaseModel:
        if binary_idl_object.tag == MESSAGEPACK:
            dict_obj = msgpack.loads(binary_idl_object.value)
            python_val = expected_python_type.model_validate(obj=dict_obj, strict=False)
            return python_val
 ```

Future-Outlier replied:
If we can guarantee the console can generate an integer but not float from the input, then we can delete it.

eapolinario replied:
how is this going to work though? Do we also do a version check of the backend?

wild-endeavor replied:
After console does the right thing, won't this value be coming in through the binary value instead?  Instead of `lv.scalar.primitive.integer/float`.

Future-Outlier commented:
@lukas503 
Hi, I saw you add an emoji to this PR!
Do you want to help me test this out?
Search "How to test it by others?" will have a guide for you!

lukas503 commented:
Hi @Future-Outlier,

Thanks for adding the Pydantic `TypeTransformer`! Which `"How to test it by others?"` guide are you referring to?

I've been testing the code and wondered about a behavior related to caching. Specifically, I’m curious if `model_json_schema` is considered in the hash used for caching.

Here’s an example:

```python
from flytekit import task, workflow
from pydantic import BaseModel

class Config(BaseModel):
    x: int = 1
    # y: int = 4

@task(cache=True, cache_version="v1")
def task1(val: int) -> Config:
    return Config()

@task(cache=True, cache_version="v1")
def task2(cfg: Config) -> Config:
    print("CALLED!", cfg)
    return cfg

@workflow
def my_workflow():
    config = task1(val=5)
    task2(cfg=config)

if __name__ == "__main__":
    print(Config.model_json_schema())
    my_workflow()

```

When I run the workflow for the first time, nothing is cached. On the second run, the results are cached, as expected. However, if I uncomment `y: int = 4`, the tasks still remain cached. I would assume that this schema change would trigger a cache bust and re-execute the tasks. This causes failure if I update the attributes and the cache_version of task2.

Is this the expected behavior? Shouldn't schema changes like this invalidate the cache?

Future-Outlier commented:
> Hi @Future-Outlier,
> 
> Thanks for working on the Pydantic `TypeTransformer`! Which `"How to test it by others?"` guide are you referring to?
> 
> I've been testing the code locally and wondered about a behavior related to caching. Specifically, I’m curious if `model_json_schema` is considered in the hash used for caching.
> 
> Here’s an example:
> 
> ```python
> from flytekit import task, workflow
> from pydantic import BaseModel
> 
> class Config(BaseModel):
>     x: int = 1
>     # y: int = 4
> 
> @task(cache=True, cache_version="v1")
> def task1(val: int) -> Config:
>     return Config()
> 
> @task(cache=True, cache_version="v1")
> def task2(cfg: Config) -> Config:
>     print("CALLED!", cfg)
>     return cfg
> 
> @workflow
> def my_workflow():
>     config = task1(val=5)
>     task2(cfg=config)
> 
> if __name__ == "__main__":
>     print(Config.model_json_schema())
>     my_workflow()
> ```
> 
> When I run the workflow for the first time, nothing is cached. On the second run, the results are cached, as expected. However, if I uncomment `y: int = 4`, the tasks still remain cached. I would assume that this schema change would trigger a cache bust and re-execute the tasks. This causes failure if I update the attributes and the cache_version of task2.
> 
> Is this the expected behavior? Shouldn't schema changes like this invalidate the cache?

good question, will test this out and ask other maintainers if I don't know what happened, thank you <3

Future-Outlier commented:
> Hi @Future-Outlier,
> 
> Thanks for working on the Pydantic `TypeTransformer`! Which `"How to test it by others?"` guide are you referring to?
> 
> I've been testing the code locally and wondered about a behavior related to caching. Specifically, I’m curious if `model_json_schema` is considered in the hash used for caching.
> 
> Here’s an example:
> 
> ```python
> from flytekit import task, workflow
> from pydantic import BaseModel
> 
> class Config(BaseModel):
>     x: int = 1
>     # y: int = 4
> 
> @task(cache=True, cache_version="v1")
> def task1(val: int) -> Config:
>     return Config()
> 
> @task(cache=True, cache_version="v1")
> def task2(cfg: Config) -> Config:
>     print("CALLED!", cfg)
>     return cfg
> 
> @workflow
> def my_workflow():
>     config = task1(val=5)
>     task2(cfg=config)
> 
> if __name__ == "__main__":
>     print(Config.model_json_schema())
>     my_workflow()
> ```
> 
> When I run the workflow for the first time, nothing is cached. On the second run, the results are cached, as expected. However, if I uncomment `y: int = 4`, the tasks still remain cached. I would assume that this schema change would trigger a cache bust and re-execute the tasks. This causes failure if I update the attributes and the cache_version of task2.
> 
> Is this the expected behavior? Shouldn't schema changes like this invalidate the cache?

@lukas503 
sorry can you try again?
I've updated the above description.

Future-Outlier commented on flytekit/types/file/file.py:
```diff
@@ -549,12 +552,42 @@ def from_binary_idl(
         # Handle dataclass attribute access
-        if lv.scalar and lv.scalar.binary:
-            return self.from_binary_idl(lv.scalar.binary, expected_python_type)
+        if lv.scalar:
+            if lv.scalar.binary:
+                return self.from_binary_idl(lv.scalar.binary, expected_python_type)
+            if lv.scalar.generic:
+                return self.from_generic_idl(lv.scalar.generic, expected_python_type)
```
```python
class DC(BaseModel):
    ff: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))

@task(container_image=image)
def t_args(dc: DC) -> DC:
    with open(dc.ff, "r") as f:
        print(f.read())
    return dc

@task(container_image=image)
def t_ff(ff: FlyteFile) -> FlyteFile:
    with open(ff, "r") as f:
        print(f.read())
    return ff

@workflow
def wf(dc: DC) -> DC:
    t_ff(dc.ff)
    return t_args(dc=dc)
 ```
 
this is for this case input from `flyteconsole`.

lukas503 commented:
> sorry can you try again?
> I've updated the above description.

Thanks for updating the PR. I now understand the underlying issue better. It appears the caching mechanism is ignoring the output types/schema. What’s unclear to me is why the output types/schema aren’t factored into the hash used for caching. In my opinion, any interface change could invalidate the cache even the outputs. I don’t see how the old cached outputs can remain valid after an interface change.

That said, this concern isn’t directly related to the current PR, so feel free to proceed as is.

Future-Outlier commented on flytekit/interaction/click_types.py:
```diff
@@ -36,7 +36,10 @@ def is_pydantic_basemodel(python_type: typing.Type) -> bool:
     else:
         try:
-            from pydantic.v1 import BaseModel
+            from pydantic import BaseModel as BaseModelV2
+            from pydantic.v1 import BaseModel as BaseModelV1
+
+            return issubclass(python_type, BaseModelV1) or issubclass(python_type, BaseModelV2)
```
for backward compatible

Future-Outlier commented on flytekit/types/directory/__init__.py:
```diff
@@ -10,13 +10,14 @@
 
    FlyteDirectory
+   FlyteDirToMultipartBlobTransformer
    TensorboardLogs
    TFRecordsDirectory
 """
 
 import typing
 
-from .types import FlyteDirectory
+from .types import FlyteDirectory, FlyteDirToMultipartBlobTransformer
```
to import `FlyteDirToMultipartBlobTransformer` in the pydantic plugin, we have to import here.

Future-Outlier commented on flytekit/types/directory/types.py:
```diff
@@ -535,13 +538,45 @@ def from_binary_idl(
         else:
             raise TypeTransformerFailedError(f"Unsupported binary format: `{binary_idl_object.tag}`")
 
+    def from_generic_idl(self, generic: Struct, expected_python_type: typing.Type[FlyteDirectory]) -> FlyteDirectory:
+        json_str = _json_format.MessageToJson(generic)
+        python_val = json.loads(json_str)
+        path = python_val.get("path", None)
+
+        if path is None:
+            raise ValueError("FlyteDirectory's path should not be None")
+
+        return FlyteDirToMultipartBlobTransformer().to_python_value(
+            FlyteContextManager.current_context(),
+            Literal(
+                scalar=Scalar(
+                    blob=Blob(
+                        metadata=BlobMetadata(
+                            type=_core_types.BlobType(
+                                format="", dimensionality=_core_types.BlobType.BlobDimensionality.MULTIPART
+                            )
+                        ),
+                        uri=path,
+                    )
+                )
+            ),
+            expected_python_type,
+        )
```
This is literally the same as `_deserialize` functinon for Structured Dataset

wild-endeavor commented on plugins/flytekit-pydantic-v2/flytekitplugins/pydantic/transformer.py:
```diff
@@ -0,0 +1,74 @@
+        python_type: Type[BaseModel],
+        expected: types.LiteralType,
+    ) -> Literal:
+        json_str = python_val.model_dump_json()
```
Can we add a comment here saying that for pydantic, we have to go through json first?  Just to make it clear.

Future-Outlier replied:
YES of course

wild-endeavor commented on plugins/flytekit-pydantic-v2/flytekitplugins/pydantic/__init__.py:
```diff
@@ -0,0 +1,25 @@
+    serialize_structured_dataset,
+)
+
+setattr(FlyteFile, "serialize_flyte_file", serialize_flyte_file)
```
🙈 i assume we're okay with this?  Any other approach? @pingsutw @eapolinario @kumare3 - what's another approach?  setters for these functions?

Future-Outlier replied:
I think this is the best way since we can't write the code like the given example in official pydantic documentation.
https://docs.pydantic.dev/latest/api/functional_validators/#pydantic.functional_validators.model_validator

wild-endeavor commented:
@Future-Outlier there's a bunch of code in here about deserializing from proto struct... is that really necessary? I think we should plan for the front-end to be able to differentiate in the future.  thoughts @eapolinario?

Future-Outlier commented on flytekit/core/type_engine.py:
```diff
@@ -242,6 +245,13 @@ def from_binary_idl(self, binary_idl_object: Binary, expected_python_type: Type[
         else:
             raise TypeTransformerFailedError(f"Unsupported binary format `{binary_idl_object.tag}`")
 
+    def from_generic_idl(self, generic: Struct, expected_python_type: Type[T]) -> Optional[T]:
```
YES, but I don't think `add a flag to the type` is a good choice now, since it might require IDL changes, and we should be super careful about IDL changes.

Future-Outlier commented on flytekit/core/type_engine.py:
```diff
@@ -2256,6 +2301,17 @@ def _check_and_covert_float(lv: Literal) -> float:
     raise TypeTransformerFailedError(f"Cannot convert literal {lv} to float")
 
 
+def _check_and_covert_int(lv: Literal) -> int:
```
1. yes I'm going to provide a more meaningful name
2. If we use protobuf struct as input from the console, this can't be deleted.
We have to provide an identifier to the console, then the console knows it can generate msgpack idl.

Future-Outlier commented on plugins/flytekit-pydantic-v2/README.md:
```diff
@@ -0,0 +1,192 @@
+    APPROVED = "approved"
+    REJECTED = "rejected"
+
+class InnerDC(BaseModel):
```
YES, will do it today

Future-Outlier replied:
Note: we didn't support 2 level optional attr access now.

Future-Outlier commented on flytekit/types/directory/types.py:
```diff
@@ -506,42 +489,106 @@ def to_literal(
             raise TypeTransformerFailedError(f"Unsupported binary format: `{binary_idl_object.tag}`")
 
+    def from_generic_idl(self, generic: Struct, expected_python_type: typing.Type[FlyteDirectory]) -> FlyteDirectory:
+        """
```
thinking, this is a HARD question for me.

Future-Outlier replied:
I think if the console is supported, then we can remove this in the future, but this might happend in the next year...

Future-Outlier replied:
update: I think deprecated is better than remove.
Imagine our OSS user only upgrade propeller but didn't upgrade flyteconsole.
Does this make sense?

Future-Outlier commented:
> @Future-Outlier there's a bunch of code in here about deserializing from proto struct... is that really necessary? I think we should plan for the front-end to be able to differentiate in the future. thoughts @eapolinario?

YES this is necessary, I'm pretty sure.
If you want flyte console input works, then you have to support that.
```python
@dataclass
class DC:
    dict_int_ff: Dict[int, FlyteFile]

@workflow
def wf(dc: DC):
    t_ff(dc.dict_int_ff)
```

Think about what will happen in the above case.

```python
Life Cycle:
json str            -> protobuf struct         -> resolved protobuf struct   -> dictionary                -> expected Python object
(console user input)   (console output)           (propeller)                   (flytekit dict transformer)  (mashumaro decoder)
```

Future-Outlier commented:
> @Future-Outlier there's a bunch of code in here about deserializing from proto struct... is that really necessary? I think we should plan for the front-end to be able to differentiate in the future. thoughts @eapolinario?

YES, IT IS NECESSARY, or you can't support attr access.
we should definitely support this.
The use case is super important.

Read this can explain why
https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR2104-R2106

Future-Outlier commented:
TODO:
dataclass optional test
basemodel optional test
union transformer discussion
dataclass in basemodel in README.md

Future-Outlier commented on flytekit/core/type_engine.py:
```diff
@@ -1866,6 +1877,9 @@ async def async_to_python_value(
         # This is serial, not actually async, but should be okay since it's more reasonable for Unions.
         for v in get_args(expected_python_type):
             try:
+                if found_res and v is NoneType:
+                    continue
+
```
When the input is a msgpack idl, and the expected python type is an Optional.
`trans.async_to_python_value(ctx, lv, int)` and
`trans.async_to_python_value(ctx, lv, None)` both works.
This will cause ambiguity, so I use `continue` to skip this. 

For example:
```python
@dataclass 
class DC:
       a: Optional[int]

@workflow
def wf(dc: DC):
      t_optional_int(a=a)
```

Future-Outlier commented on plugins/flytekit-pydantic-v2/setup.py:
```diff
@@ -0,0 +1,40 @@
+
+microlib_name = f"flytekitplugins-{PLUGIN_NAME}-v2"
+
+plugin_requires = ["flytekit>1.13.7", "pydantic>=2.6.0"]
```
I don't think so unless pydantic v3 breaks v2.
Dicussion about pydantic v3: https://github.com/pydantic/pydantic/issues/10033

Future-Outlier commented on flytekit/core/type_engine.py:
```diff
@@ -229,16 +230,55 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
+
+            This is to avoid the above ambiguity.
+            """
+            if expected_python_type is NoneType:
+                assert python_val is None
+            else:
+                assert python_val is not None
```
Yes, let me try it

Future-Outlier replied:
We have to add these logic in the following place in the Union Transformer, which is too much and make the transformer's code harder to read.

1. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1904-R1905
2. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1909-R1910
3. https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1921-R1924

Future-Outlier replied:
And I think the current implementation follows the origin's philosophy, which means
"Handle the function `to python value` in the selected type transformer, succeed or fail is determined by it."
So that's why I do this implementation.

Related Code:

pingsutw replied:
okay, that makes sense. thank you

Future-Outlier replied:
Done, thinking how to add tests for this case, extremely important.

Future-Outlier commented on plugins/flytekit-pydantic-v2/flytekitplugins/pydantic/transformer.py:
```diff
@@ -0,0 +1,74 @@
+            return self.from_binary_idl(lv.scalar.binary, expected_python_type)  # type: ignore
+
+        json_str = _json_format.MessageToJson(lv.scalar.generic)
+        python_val = expected_python_type.model_validate_json(json_str, strict=False)
```
Let me deep dive into it TODAY, thank you!

Future-Outlier replied:
update: it's doable, but I need to change the behavior in `click_types.py`, thank you for this level of consideration!

Future-Outlier replied:
Done, thinking how to add tests for this case, is extremely important.

wild-endeavor reviewed (commented):
I think we should rip this out actually.  can we move it into core flytekit?

protect it with an import check.  move all these files out of the plugins folder, into core flytekit.  what do you think @eapolinario

Future-Outlier commented:
> I think we should rip this out actually. can we move it into core flytekit?
> 
> protect it with an import check. move all these files out of the plugins folder, into core flytekit. what do you think @eapolinario ? ketan was of the same mind.

YES IT CAN.

Future-Outlier commented:
todo: add this to flytesnack example
```md
# Flytekit Pydantic Plugin

Pydantic is a data validation and settings management library that uses Python type annotations to enforce type hints at runtime and provide user-friendly errors when data is invalid. Pydantic models are classes that inherit from `pydantic.BaseModel` and are used to define the structure and validation of data using Python type annotations.

The plugin adds type support for pydantic models.

To install the plugin, run the following command:

```bash
pip install flytekitplugins-pydantic-v2
```

## Type Example
```python
from enum import Enum
import os
from typing import Dict, List, Optional

import pandas as pd
from pydantic import BaseModel, Field

from flytekit.types.schema import FlyteSchema
from flytekit.types.structured import StructuredDataset
from flytekit.types.file import FlyteFile
from flytekit.types.directory import FlyteDirectory
from flytekit import task, workflow, ImageSpec


image = ImageSpec(packages=["flytekitplugins-pydantic-v2",
                            "pandas",
                            "pyarrow"],
                            registry="localhost:30000",
                            )

class Status(Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"

class InnerBM(BaseModel):
    a: int = -1
    b: float = 2.1
    c: str = "Hello, Flyte"
    d: bool = False
    e: List[int] = Field(default_factory=lambda: [0, 1, 2, -1, -2])
    f: List[FlyteFile] = Field(default_factory=lambda: [FlyteFile("s3://my-s3-bucket/example.txt")])
    g: List[List[int]] = Field(default_factory=lambda: [[0], [1], [-1]])
    h: List[Dict[int, bool]] = Field(default_factory=lambda: [{0: False}, {1: True}, {-1: True}])
    i: Dict[int, bool] = Field(default_factory=lambda: {0: False, 1: True, -1: False})
    j: Dict[int, FlyteFile] = Field(default_factory=lambda: {0: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             1: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             -1: FlyteFile("s3://my-s3-bucket/example.txt")})
    k: Dict[int, List[int]] = Field(default_factory=lambda: {0: [0, 1, -1]})
    l: Dict[int, Dict[int, int]] = Field(default_factory=lambda: {1: {-1: 0}})
    m: dict = Field(default_factory=lambda: {"key": "value"})
    n: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))
    o: FlyteDirectory = Field(default_factory=lambda: FlyteDirectory("s3://my-s3-bucket/s3_flyte_dir"))
    enum_status: Status = Status.PENDING
    sd: StructuredDataset = Field(default_factory=lambda: StructuredDataset(
        uri="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/7f31035fdf92510e40ee9340f9e5bf34",
        file_format="parquet"))
    fsc: FlyteSchema = Field(default_factory=lambda: FlyteSchema(
        remote_path="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/ab3aef21302d0529daef8c43825c3fdf"))

class BM(BaseModel):
    a: int = -1
    b: float = 2.1
    c: str = "Hello, Flyte"
    d: bool = False
    e: List[int] = Field(default_factory=lambda: [0, 1, 2, -1, -2])
    f: List[FlyteFile] = Field(default_factory=lambda: [FlyteFile("s3://my-s3-bucket/example.txt")])
    g: List[List[int]] = Field(default_factory=lambda: [[0], [1], [-1]])
    h: List[Dict[int, bool]] = Field(default_factory=lambda: [{0: False}, {1: True}, {-1: True}])
    i: Dict[int, bool] = Field(default_factory=lambda: {0: False, 1: True, -1: False})
    j: Dict[int, FlyteFile] = Field(default_factory=lambda: {0: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             1: FlyteFile("s3://my-s3-bucket/example.txt"),
                                                             -1: FlyteFile("s3://my-s3-bucket/example.txt")})
    k: Dict[int, List[int]] = Field(default_factory=lambda: {0: [0, 1, -1]})
    l: Dict[int, Dict[int, int]] = Field(default_factory=lambda: {1: {-1: 0}})
    m: dict = Field(default_factory=lambda: {"key": "value"})
    n: FlyteFile = Field(default_factory=lambda: FlyteFile("s3://my-s3-bucket/example.txt"))
    o: FlyteDirectory = Field(default_factory=lambda: FlyteDirectory("s3://my-s3-bucket/s3_flyte_dir"))
    inner_dc: InnerBM = Field(default_factory=lambda: InnerBM())
    enum_status: Status = Status.PENDING
    sd: StructuredDataset = Field(default_factory=lambda: StructuredDataset(
        uri="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/7f31035fdf92510e40ee9340f9e5bf34",
        file_format="parquet"))
    fsc: FlyteSchema = Field(default_factory=lambda: FlyteSchema(remote_path="s3://my-s3-bucket/data/uk/ahlg7qw7q5m4np7vwdqm-n0-0/ab3aef21302d0529daef8c43825c3fdf"))

@task(container_image=image)
def t_dc(dc: BM) -> BM:
    return dc

@task(container_image=image)
def t_inner(inner_dc: InnerBM):
    assert isinstance(inner_dc, InnerBM)

    expected_file_content = "Default content"

    # f: List[FlyteFile]
    for ff in inner_dc.f:
        assert isinstance(ff, FlyteFile)
        with open(ff, "r") as f:
            assert f.read() == expected_file_content
    # j: Dict[int, FlyteFile]
    for _, ff in inner_dc.j.items():
        assert isinstance(ff, FlyteFile)
        with open(ff, "r") as f:
            assert f.read() == expected_file_content
    # n: FlyteFile
    assert isinstance(inner_dc.n, FlyteFile)
    with open(inner_dc.n, "r") as f:
        assert f.read() == expected_file_content
    # o: FlyteDirectory
    assert isinstance(inner_dc.o, FlyteDirectory)
    assert not inner_dc.o.downloaded
    with open(os.path.join(inner_dc.o, "example.txt"), "r") as fh:
        assert fh.read() == expected_file_content
    assert inner_dc.o.downloaded
    print("Test InnerBM Successfully Passed")
    # enum: Status
    assert inner_dc.enum_status == Status.PENDING


@task(container_image=image)
def t_test_all_attributes(a: int, b: float, c: str, d: bool, e: List[int], f: List[FlyteFile], g: List[List[int]],
                          h: List[Dict[int, bool]], i: Dict[int, bool], j: Dict[int, FlyteFile],
                          k: Dict[int, List[int]], l: Dict[int, Dict[int, int]], m: dict,
                          n: FlyteFile, o: FlyteDirectory,
                          enum_status: Status,
                          sd: StructuredDataset,
                          fsc: FlyteSchema,
                          ):
    # Strict type checks for simple types
    assert isinstance(a, int), f"a is not int, it's {type(a)}"
    assert a == -1
    assert isinstance(b, float), f"b is not float, it's {type(b)}"
    assert isinstance(c, str), f"c is not str, it's {type(c)}"
    assert isinstance(d, bool), f"d is not bool, it's {type(d)}"

    # Strict type checks for List[int]
    assert isinstance(e, list) and all(isinstance(i, int) for i in e), "e is not List[int]"

    # Strict type checks for List[FlyteFile]
    assert isinstance(f, list) and all(isinstance(i, FlyteFile) for i in f), "f is not List[FlyteFile]"

    # Strict type checks for List[List[int]]
    assert isinstance(g, list) and all(
        isinstance(i, list) and all(isinstance(j, int) for j in i) for i in g), "g is not List[List[int]]"

    # Strict type checks for List[Dict[int, bool]]
    assert isinstance(h, list) and all(
        isinstance(i, dict) and all(isinstance(k, int) and isinstance(v, bool) for k, v in i.items()) for i in h
    ), "h is not List[Dict[int, bool]]"

    # Strict type checks for Dict[int, bool]
    assert isinstance(i, dict) and all(
        isinstance(k, int) and isinstance(v, bool) for k, v in i.items()), "i is not Dict[int, bool]"

    # Strict type checks for Dict[int, FlyteFile]
    assert isinstance(j, dict) and all(
        isinstance(k, int) and isinstance(v, FlyteFile) for k, v in j.items()), "j is not Dict[int, FlyteFile]"

    # Strict type checks for Dict[int, List[int]]
    assert isinstance(k, dict) and all(
        isinstance(k, int) and isinstance(v, list) and all(isinstance(i, int) for i in v) for k, v in
        k.items()), "k is not Dict[int, List[int]]"

    # Strict type checks for Dict[int, Dict[int, int]]
    assert isinstance(l, dict) and all(
        isinstance(k, int) and isinstance(v, dict) and all(
            isinstance(sub_k, int) and isinstance(sub_v, int) for sub_k, sub_v in v.items())
        for k, v in l.items()), "l is not Dict[int, Dict[int, int]]"

    # Strict type check for a generic dict
    assert isinstance(m, dict), "m is not dict"

    # Strict type check for FlyteFile
    assert isinstance(n, FlyteFile), "n is not FlyteFile"

    # Strict type check for FlyteDirectory
    assert isinstance(o, FlyteDirectory), "o is not FlyteDirectory"

    # # Strict type check for Enum
    assert isinstance(enum_status, Status), "enum_status is not Status"

    assert isinstance(sd, StructuredDataset), "sd is not StructuredDataset"
    print("sd:", sd.open(pd.DataFrame).all())

    assert isinstance(fsc, FlyteSchema), "fsc is not FlyteSchema"
    print("fsc: ", fsc.open().all())

    print("All attributes passed strict type checks.")


@workflow
def wf(dc: BM):
    t_dc(dc=dc)
    t_inner(inner_dc=dc.inner_dc)
    t_test_all_attributes(a=dc.a, b=dc.b, c=dc.c,
                          d=dc.d, e=dc.e, f=dc.f,
                          g=dc.g, h=dc.h, i=dc.i,
                          j=dc.j, k=dc.k, l=dc.l,
                          m=dc.m, n=dc.n, o=dc.o,
                          enum_status=dc.enum_status,
                          sd=dc.sd,
                          fsc=dc.fsc,
                          )

    t_test_all_attributes(a=dc.inner_dc.a, b=dc.inner_dc.b, c=dc.inner_dc.c,
                          d=dc.inner_dc.d, e=dc.inner_dc.e, f=dc.inner_dc.f,
                          g=dc.inner_dc.g, h=dc.inner_dc.h, i=dc.inner_dc.i,
                          j=dc.inner_dc.j, k=dc.inner_dc.k, l=dc.inner_dc.l,
                          m=dc.inner_dc.m, n=dc.inner_dc.n, o=dc.inner_dc.o,
                          enum_status=dc.inner_dc.enum_status,
                          sd=dc.inner_dc.sd,
                          fsc=dc.inner_dc.fsc,
                          )
```

```

Future-Outlier commented on flytekit/__init__.py:
```diff
@@ -243,6 +244,7 @@
 from flytekit.core.workflow import ImperativeWorkflow as Workflow
 from flytekit.core.workflow import WorkflowFailurePolicy, reference_workflow, workflow
 from flytekit.deck import Deck
+from flytekit.extras import pydantic
```
If we move to lazy import transformer, this will fail to have custom serialize and deserialize behavior, still investigation the root cause.

Future-Outlier reviewed (commented):
Just tested it and it works
<img width="1181" alt="image" src="https://github.com/user-attachments/assets/30705407-8995-4f57-8b2d-1dc3c404f283">

eapolinario commented on .github/workflows/pythonbuild.yml:
```diff
@@ -356,7 +356,8 @@ jobs:
           - flytekit-papermill
           - flytekit-polars
-          - flytekit-pydantic
+          # TODO: remove pydantic v1 plugin, since v2 is in core already
+          # flytekit-pydantic
```
Can we remove it?

eapolinario commented on flytekit/core/type_engine.py:
```diff
@@ -215,16 +215,32 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
 
     def from_binary_idl(self, binary_idl_object: Binary, expected_python_type: Type[T]) -> Optional[T]:
+        """
+        TODO: Add more comments to explain the lifecycle of attribute access.
```
fill in TODO

eapolinario commented on plugins/flytekit-pydantic/setup.py:
```diff
@@ -4,7 +4,7 @@
 microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
 
-plugin_requires = ["flytekit>=1.7.0b0", "pydantic"]
+plugin_requires = ["flytekit>=1.7.0b0", "pydantic<2"]
```
Can we also leave a warning in the [README.md](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-pydantic/README.md) explaining that we're deprecating this plugin?

Future-Outlier replied:
@eapolinario 
> how is this going to work though? Do we also do a version check of the backend?
No this is just for supporting the case I've mentioned above.
we didn't support this before and I think we should do it.

eapolinario reviewed (commented):
Just a few minor things, otherwise, it's looking pretty good.

Future-Outlier commented on flytekit/extras/pydantic/decorator.py:
```diff
@@ -0,0 +1,57 @@
+    from typing_extensions import Literal as typing_literal
+
+    def model_serializer(
+        __f: Union[Callable[..., Any], None] = None,
+        *,
+        mode: typing_literal["plain", "wrap"] = "plain",
+        when_used: typing_literal["always", "unless-none", "json", "json-unless-none"] = "always",
+        return_type: Any = None,
+    ) -> Callable[[Any], Any]:
+        """Placeholder decorator for Pydantic model_serializer."""
+
+        def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+            def wrapper(*args, **kwargs):
+                raise Exception(
+                    "Pydantic is not installed.\n" "Please install Pydantic version > 2 to use this feature."
+                )
+
+            return wrapper
+
+        # If no function (__f) is provided, return the decorator
+        if __f is None:
+            return decorator
+        # If __f is provided, directly decorate the function
+        return decorator(__f)
+
+    def model_validator(
+        *,
+        mode: typing_literal["wrap", "before", "after"],
+    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+        """Placeholder decorator for Pydantic model_validator."""
+
+        def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+            def wrapper(*args, **kwargs):
+                raise Exception(
+                    "Pydantic is not installed.\n" "Please install Pydantic version > 2 to use this feature."
+                )
+
+            return wrapper
+
+        return decorator
```
we want to make here to work.
This syntax is more readable then `setattr`
https://github.com/flyteorg/flytekit/pull/2792/files#diff-22cf9c7153b54371b4a77331ddf276a082cf4b3c5e7bd1595dd67232288594fdR168-R176

wild-endeavor replied:
it's to support the case where pydantic is not installed at all.  cuz yeah it looks nicer in the real File/Directory class, but we also want it to not fail ofc.

Future-Outlier commented:
Let's merge it.
cc @eapolinario @wild-endeavor @pingsutw

wild-endeavor commented on flytekit/core/type_engine.py:
```diff
@@ -2194,6 +2304,34 @@ def _check_and_covert_float(lv: Literal) -> float:
 
+def _handle_flyte_console_float_input_to_int(lv: Literal) -> int:
+    """
+    Flyte Console is written by JavaScript and JavaScript has only one number type which is float.
```
technically javascript's number type is [Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number) but yeah, sometimes it keeps track of trailing 0s and sometimes it doesn't.

Future-Outlier replied:
thank you, it's important to make words more accurately

wild-endeavor commented on flytekit/extras/pydantic/transformer.py:
```diff
@@ -0,0 +1,81 @@
+        super().__init__("Pydantic Transformer", BaseModel, enable_type_assertions=False)
+
+    def get_literal_type(self, t: Type[BaseModel]) -> LiteralType:
+        schema = t.model_json_schema()
```
In a future PR, can we add some unit tests to ensure that we're correctly extracting [default values](https://github.com/flyteorg/flyte/issues/5887#issuecomment-2436565600) into the schema?

Future-Outlier replied:
Yes we can

Future-Outlier replied:
I'll ask more question about this unit tests PR if not totally understand, thank you

wild-endeavor commented on tests/flytekit/unit/extras/pydantic/test_pydantic_transformer.py:
```diff
@@ -0,0 +1,775 @@
+
+    bm = BM()
+    wf(bm=bm)
+
```
can you add two new lines between tests?  I know it doesn't matter, but pycharm complains.

wild-endeavor commented on tests/flytekit/unit/extras/pydantic/test_pydantic_transformer.py:
```diff
@@ -0,0 +1,775 @@
+
+def test_flytetypes_in_pydantic_basemodel_wf(local_dummy_file, local_dummy_directory):
+    class InnerBM(BaseModel):
+        flytefile: FlyteFile = field(default_factory=lambda: FlyteFile(local_dummy_file))
```
little bit confused about pydantic here.  Are you supposed to use `dataclasses.field` here instead of `pydantic.Field`?

Future-Outlier replied:
both are supported, this is a good catch, should tell users `field from dataclass and Field from pydantic` are supported.

wild-endeavor commented on tests/flytekit/unit/types/structured_dataset/test_structured_dataset.py:
```diff
@@ -49,16 +48,13 @@
 
 def generate_pandas() -> pd.DataFrame:
     return pd.DataFrame({"name": ["Tom", "Joseph"], "age": [20, 22]})
-
```
keep spaces plz

wild-endeavor commented on tests/flytekit/unit/extras/pydantic/test_pydantic_transformer.py:
```diff
@@ -0,0 +1,775 @@
+        flytefile: FlyteFile = field(default_factory=lambda: FlyteFile(local_dummy_file))
+        flytedir: FlyteDirectory = field(default_factory=lambda: FlyteDirectory(local_dummy_directory))
+
+    class BM(BaseModel):
```
maybe in a different file, but can we add pydantic models that contain `dataclass` and also `dataclass`es that contain Pydantic models?  I know some people have been asking for that, be good to have some tests for it.

Thank you!

Future-Outlier replied:
> pydantic models that contain dataclass
this is supported

> dataclasses that contain Pydantic models
this is never supported, if people really want it, we can do it in a seperate PR

Future-Outlier replied:
I'll add unit tests today, thank you so much

wild-endeavor reviewed (commented):
🙇 thank you @Future-Outlier 🙏  this is going to be fantastic.

Future-Outlier commented:
cc @wild-endeavor 
> After console does the right thing, won't this value be coming in through the binary value instead? Instead of lv.scalar.primitive.integer/float.

if console can turn input to expected python type (int or float), then yes.

Future-Outlier commented:
cc @wild-endeavor @eapolinario @pingsutw 

### Summary of Changes:

1. **Linting**: Applied linting to all test files in the final commit.
2. **Folder Renaming**: Changed the folder name from `flytekit.extras.pydantic` to `flytekit.extras.pydantic_transformer`.
https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR1197-R1198
3. **JavaScript Number Handling**: Updated comments related to JavaScript's number handling.
https://github.com/flyteorg/flytekit/pull/2792/files#diff-68624e201f644896b015de1a5716eabef9b8b5a106caba1d9f21c51c0708684bR2308-R2312
4. **Tests Added**:
   - Added tests using both `dataclass.field` and `pydantic.Field`.
   - Created unit tests for Pydantic models inside dataclasses (note: this will raise an error, and an issue will be created).
   https://github.com/flyteorg/flytekit/pull/2792/files#diff-779944519da7ce37ef29492b53edb53d653884263629341e4721c22d2223dedd
   - Created unit tests for dataclasses inside Pydantic models (using `field`).
   https://github.com/flyteorg/flytekit/pull/2792/files#diff-23d316d571d3169735f378aa89e677ee2d30f0d2b6c9004d08341180aa0d5e3a
5. **Decorator Comments**: Added comments in the Pydantic transformer decorator for scenarios where Pydantic is not installed.
https://github.com/flyteorg/flytekit/pull/2792/files#diff-65db41859ac618afd17929975a6975810003ee243e2d88f5898ba49667524e27R12-R15

### Notes:

1. I used `autopep8` as the linter for some `test_*.py` files, which resulted in around 1000 lines of code changes (please ignore these changes).
2. I encountered an issue while rebasing with the master branch and failed to pull others' commits into this PR (my mistake). Consequently, I had to manually set the DCO to pass. Apologies for the inconvenience.

wild-endeavor commented:
> if people really want it, we can do it in a seperate PR
yeah, people really want it.  let's do it in a separate PR.
